src/coap_cache.c

```diff
@@ -0,0 +1,245 @@
+/* coap_cache.c -- Caching of CoAP requests
+*
+* Copyright (C) 2020 Olaf Bergmann <bergmann@tzi.org>
+*
+* This file is part of the CoAP library libcoap. Please see
+* README for terms of use.
+*/
+
+#include "coap_internal.h"
+
+/* Determines if the given option_type denotes an option type that can
+ * be used as CacheKey. Options that can be cache keys are not Unsafe
+ * and not marked explicitly as NoCacheKey. */
+static int
+is_cache_key(const coap_context_t *ctx, uint16_t option_type) {
+  int nocachekey = (option_type & 0x1e) == 0x1c;
+  size_t i;
+
+  /*
+   * https://tools.ietf.org/html/rfc7641#section-2 Observe is not a
+   * part of the cache-key.
+   */
+  if (option_type == COAP_OPTION_OBSERVE) {
+    nocachekey = 1;
+  }
+
+  /* Check for option user has defined as not part of cache-key */
+  for (i = 0; i < ctx->cache_ignore_count; i++) {
+    if (ctx->cache_ignore_options[i] == option_type) {
+      nocachekey = 1;
+      break;
+    }
+  }
+
+  return nocachekey == 0;
+}
+
+int
+coap_cache_ignore_options(coap_context_t *ctx,
+                          const uint16_t *options,
+                          size_t count) {
+  if (ctx->cache_ignore_options) {
+    coap_free(ctx->cache_ignore_options);
+  }
+  ctx->cache_ignore_options = coap_malloc(count * sizeof(options[0]));
```

Comment:
Depending on the actual implementation of `coap_malloc()` passing `0` as size might be problematic. I suggest special handling for the case `count == 0` (also in order to avoid the warning when `coap_cache_ignore_options(ctx, NULL, 0)` is called to clear the list of ignored options).

Reply:
Special case added in.

Reply:
I think that `ctx->cache_ignore_options` must be set to NULL in the else-case (`count == 0`) or right after the `coap_free()` to avoid a double free.

Reply:
Good catch - fixed.